File: examples/python/jwk_v1_query_messages.py
#!/usr/bin/env python3
"""
xion.jwk.v1 Query Messages — Protobuf Examples

This file demonstrates how to construct, encode, and decode
query messages from the JWK Module (xion.jwk.v1).

Queries covered:
- QueryAudience: Get audience by ID
- QueryAudienceClaim: Get specific audience claim
- QueryAudienceClaims: List all claims for an audience
- QueryAudiences: List all registered audiences
- QueryParams: Get module parameters
"""

import sys
import os

# Add the python types to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../python/types'))

from xion.jwk.v1 import query_pb2 as jwk_query
from google.protobuf.json_format import MessageToJson


# =============================================================================
# QueryAudienceRequest / Response
# =============================================================================
# Get information about a specific audience.
# Type URL: /xion.jwk.v1.QueryAudienceRequest

def query_audience_examples():
    print('=== QueryAudience Examples ===\n')

    # 1. Construct request
    request = jwk_query.QueryAudienceRequest()
    request.aud = 'your-app-audience-id'

    print('1. Constructed QueryAudienceRequest:')
    print(f'   aud: {request.aud}')
    print(f'   typeUrl: /xion.jwk.v1.QueryAudienceRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = jwk_query.QueryAudienceResponse()
    response.audience.aud = 'your-app-audience-id'
    response.audience.key = 'https://your-issuer.com/.well-known/jwks.json'
    response.audience.admin = 'xion1admin...'

    print('3. Constructed Response:')
    print(f'   audience.aud: {response.audience.aud}')
    print(f'   audience.key: {response.audience.key}')
    print(f'   audience.admin: {response.audience.admin}')
    print(f'   typeUrl: /xion.jwk.v1.QueryAudienceResponse')

    # 4. JSON format
    json_str = MessageToJson(response)
    print(f'\n4. JSON format:')
    print(f'   {json_str}')

    return request, response


# =============================================================================
# QueryAudienceClaimRequest / Response
# =============================================================================
# Get a specific claim for an audience.
# Type URL: /xion.jwk.v1.QueryAudienceClaimRequest

def query_audience_claim_examples():
    print('\n=== QueryAudienceClaim Examples ===\n')

    # 1. Construct request
    request = jwk_query.QueryAudienceClaimRequest()
    request.aud_hash = b'audience-hash-bytes'
    request.claim_hash = b'claim-hash-bytes'

    print('1. Constructed QueryAudienceClaimRequest:')
    print(f'   aud_hash: {request.aud_hash.hex()}')
    print(f'   claim_hash: {request.claim_hash.hex()}')
    print(f'   typeUrl: /xion.jwk.v1.QueryAudienceClaimRequest')

    # 2. Encode
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = jwk_query.QueryAudienceClaimResponse()
    response.claim.aud_hash = b'audience-hash-bytes'
    response.claim.claim_hash = b'claim-hash-bytes'

    print('3. Constructed Response:')
    print(f'   claim.aud_hash: {response.claim.aud_hash.hex()}')
    print(f'   claim.claim_hash: {response.claim.claim_hash.hex()}')

    return request, response


# =============================================================================
# QueryAudienceClaimsRequest / Response
# =============================================================================
# List all claims for a specific audience.
# Type URL: /xion.jwk.v1.QueryAudienceClaimsRequest

def query_audience_claims_examples():
    print('\n=== QueryAudienceClaims Examples ===\n')

    # 1. Construct request with pagination
    request = jwk_query.QueryAudienceClaimsRequest()
    request.aud_hash = b'audience-hash-bytes'
    request.pagination.limit = 10
    request.pagination.offset = 0

    print('1. Constructed QueryAudienceClaimsRequest:')
    print(f'   aud_hash: {request.aud_hash.hex()}')
    print(f'   pagination.limit: {request.pagination.limit}')
    print(f'   typeUrl: /xion.jwk.v1.QueryAudienceClaimsRequest')

    # 2. Encode
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = jwk_query.QueryAudienceClaimsResponse()
    claim1 = response.claims.add()
    claim1.aud_hash = b'audience-hash-bytes'
    claim1.claim_hash = b'claim-hash-1'

    claim2 = response.claims.add()
    claim2.aud_hash = b'audience-hash-bytes'
    claim2.claim_hash = b'claim-hash-2'

    print('3. Constructed Response:')
    print(f'   claims count: {len(response.claims)}')
    for i, claim in enumerate(response.claims):
        print(f'   [{i}] claim_hash: {claim.claim_hash.hex()}')

    return request, response


# =============================================================================
# QueryAudiencesRequest / Response
# =============================================================================
# List all registered audiences.
# Type URL: /xion.jwk.v1.QueryAudiencesRequest

def query_audiences_examples():
    print('\n=== QueryAudiences Examples ===\n')

    # 1. Construct request with pagination
    request = jwk_query.QueryAudiencesRequest()
    request.pagination.limit = 10

    print('1. Constructed QueryAudiencesRequest:')
    print(f'   pagination.limit: {request.pagination.limit}')
    print(f'   typeUrl: /xion.jwk.v1.QueryAudiencesRequest')

    # 2. Encode
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = jwk_query.QueryAudiencesResponse()

    aud1 = response.audiences.add()
    aud1.aud = 'app-audience-1'
    aud1.key = 'https://issuer1.com/.well-known/jwks.json'
    aud1.admin = 'xion1admin1...'

    aud2 = response.audiences.add()
    aud2.aud = 'app-audience-2'
    aud2.key = 'https://issuer2.com/.well-known/jwks.json'
    aud2.admin = 'xion1admin2...'

    print('3. Constructed Response:')
    print(f'   audiences count: {len(response.audiences)}')
    for aud in response.audiences:
        print(f'     - aud: {aud.aud}')
        print(f'       admin: {aud.admin}')

    return request, response


# =============================================================================
# QueryParamsRequest / Response
# =============================================================================
# Get JWK module parameters.
# Type URL: /xion.jwk.v1.QueryParamsRequest

def query_params_examples():
    print('\n=== QueryParams Examples ===\n')

    # 1. Construct request (no parameters)
    request = jwk_query.QueryParamsRequest()

    print('1. Constructed QueryParamsRequest:')
    print('   (no parameters)')
    print(f'   typeUrl: /xion.jwk.v1.QueryParamsRequest')

    # 2. Encode
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = jwk_query.QueryParamsResponse()
    # params would contain module configuration
    # response.params.some_param = value

    print('3. Constructed Response:')
    print('   (params would contain module configuration)')
    print(f'   typeUrl: /xion.jwk.v1.QueryParamsResponse')

    return request, response


# =============================================================================
# Run All Examples
# =============================================================================

def main():
    print('╔════════════════════════════════════════════════════════════╗')
    print('║  xion.jwk.v1 Query Messages — Protobuf Examples            ║')
    print('╚════════════════════════════════════════════════════════════╝\n')

    query_audience_examples()
    query_audience_claim_examples()
    query_audience_claims_examples()
    query_audiences_examples()
    query_params_examples()

    print('\n═══════════════════════════════════════════════════════════════')
    print('Type URL Reference (Requests):')
    print('  QueryAudienceRequest:       /xion.jwk.v1.QueryAudienceRequest')
    print('  QueryAudienceClaimRequest:  /xion.jwk.v1.QueryAudienceClaimRequest')
    print('  QueryAudienceClaimsRequest: /xion.jwk.v1.QueryAudienceClaimsRequest')
    print('  QueryAudiencesRequest:      /xion.jwk.v1.QueryAudiencesRequest')
    print('  QueryParamsRequest:         /xion.jwk.v1.QueryParamsRequest')
    print('\nNote: JWK module manages JWT audiences for social login')
    print('and meta-transaction authentication in abstract accounts.')
    print('═══════════════════════════════════════════════════════════════')


if __name__ == '__main__':
    main()

File: examples/rust/examples/jwk_v1_query_messages.rs
//! xion.jwk.v1 Query Messages — Protobuf Examples
//!
//! This file demonstrates how to construct, encode, and decode
//! query messages from the JSON Web Key Module (xion.jwk.v1).
//!
//! Queries covered:
//! - QueryParams: Get module parameters (time_offset, deployment_gas)
//! - QueryAudience: Get a single audience by ID
//! - QueryAudienceAll: List all registered audiences (paginated)
//! - QueryAudienceClaim: Get an audience claim by hash
//! - QueryValidateJWT: Validate a JWT signature

use prost::Message;

use xion_types::types::cosmos_base_query_v1beta1::{PageRequest, PageResponse};
use xion_types::types::xion_jwk_v1::{
    Audience, AudienceClaim, Params, PrivateClaim, QueryAudienceAllRequest,
    QueryAudienceAllResponse, QueryAudienceClaimRequest, QueryAudienceClaimResponse,
    QueryAudienceRequest, QueryAudienceResponse, QueryParamsRequest, QueryParamsResponse,
    QueryValidateJwtRequest, QueryValidateJwtResponse,
};

// =============================================================================
// QueryParams
// =============================================================================
// Get the JWK module parameters.
// Endpoint: GET /xion/jwk/params
// Type URL: /xion.jwk.v1.QueryParamsRequest

fn query_params_examples() {
    println!("=== QueryParams Examples ===\n");

    // 1. Construct request (no parameters needed)
    let request = QueryParamsRequest {};

    println!("1. Constructed QueryParamsRequest:");
    println!("   (no parameters)");

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // === Response ===
    println!("\n--- Response ---");

    // Construct response (typically received from chain)
    let response = QueryParamsResponse {
        params: Some(Params {
            time_offset: 300000000000, // 5 minutes in nanoseconds
            deployment_gas: 100000,    // Gas required to deploy audience
        }),
    };

    println!("3. Constructed Response:");
    if let Some(ref params) = response.params {
        println!("   params.time_offset: {} ns", params.time_offset);
        println!(
            "   params.time_offset (minutes): {}",
            params.time_offset as f64 / 60_000_000_000.0
        );
        println!("   params.deployment_gas: {}", params.deployment_gas);
    }

    // 4. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response = QueryParamsResponse::decode(encoded_response.as_slice()).unwrap();
    if let Some(params) = decoded_response.params {
        println!("\n4. Decoded params.time_offset: {}", params.time_offset);
    }
}

// =============================================================================
// QueryAudience
// =============================================================================
// Get a single audience by its identifier.
// Endpoint: GET /xion/jwk/audience/{aud}
// Type URL: /xion.jwk.v1.QueryAudienceRequest

fn query_audience_examples() {
    println!("\n=== QueryAudience Examples ===\n");

    // 1. Construct request
    let request = QueryAudienceRequest {
        aud: "123456789-abc.apps.googleusercontent.com".to_string(), // OAuth client ID
    };

    println!("1. Constructed QueryAudienceRequest:");
    println!("   aud: {}", request.aud);

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryAudienceRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded aud: {}", decoded_request.aud);

    // === Response ===
    println!("\n--- Response ---");

    let response = QueryAudienceResponse {
        audience: Some(Audience {
            aud: "123456789-abc.apps.googleusercontent.com".to_string(),
            key: r#"{"kty":"RSA","n":"...","e":"AQAB"}"#.to_string(),
            admin: "xion1admin...".to_string(),
        }),
    };

    println!("4. Constructed Response:");
    if let Some(ref audience) = response.audience {
        println!("   audience.aud: {}", audience.aud);
        println!("   audience.admin: {}", audience.admin);
        println!(
            "   audience.key (truncated): {}...",
            &audience.key[..30.min(audience.key.len())]
        );
    }

    // 5. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response = QueryAudienceResponse::decode(encoded_response.as_slice()).unwrap();
    if let Some(audience) = decoded_response.audience {
        println!("\n5. Decoded audience.aud: {}", audience.aud);
    }
}

// =============================================================================
// QueryAudienceAll
// =============================================================================
// List all registered audiences with pagination.
// Endpoint: GET /xion/jwk/audience
// Type URL: /xion.jwk.v1.QueryAudienceAllRequest

fn query_audience_all_examples() {
    println!("\n=== QueryAudienceAll Examples ===\n");

    // 1. Construct request with pagination
    let request = QueryAudienceAllRequest {
        pagination: Some(PageRequest {
            key: vec![],       // Empty for first page
            offset: 0,
            limit: 10,         // Get 10 results
            count_total: true,
            reverse: false,
        }),
    };

    println!("1. Constructed QueryAudienceAllRequest:");
    if let Some(ref pagination) = request.pagination {
        println!("   pagination.limit: {}", pagination.limit);
        println!("   pagination.count_total: {}", pagination.count_total);
    }

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // === Response ===
    println!("\n--- Response ---");

    let response = QueryAudienceAllResponse {
        audience: vec![
            Audience {
                aud: "google-client-id.apps.googleusercontent.com".to_string(),
                key: r#"{"kty":"RSA","n":"..."}"#.to_string(),
                admin: "xion1admin1...".to_string(),
            },
            Audience {
                aud: "apple-client-id".to_string(),
                key: r#"{"kty":"EC","crv":"P-256",...}"#.to_string(),
                admin: "xion1admin2...".to_string(),
            },
        ],
        pagination: Some(PageResponse {
            next_key: vec![], // Empty = no more pages
            total: 2,
        }),
    };

    println!("3. Constructed Response:");
    println!("   audience count: {}", response.audience.len());
    for (i, aud) in response.audience.iter().enumerate() {
        println!("   [{}] aud: {}...", i, &aud.aud[..30.min(aud.aud.len())]);
        println!("       admin: {}", aud.admin);
    }
    if let Some(ref pagination) = response.pagination {
        println!("   pagination.total: {}", pagination.total);
    }

    // 4. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response = QueryAudienceAllResponse::decode(encoded_response.as_slice()).unwrap();
    println!("\n4. Decoded audience count: {}", decoded_response.audience.len());
}

// =============================================================================
// QueryAudienceClaim
// =============================================================================
// Get an audience claim by its hash.
// Endpoint: GET /xion/jwk/audience_claim/{hash}
// Type URL: /xion.jwk.v1.QueryAudienceClaimRequest

fn query_audience_claim_examples() {
    println!("\n=== QueryAudienceClaim Examples ===\n");

    // 1. Construct request
    let claim_hash = b"sha256-hash-of-claim".to_vec();

    let request = QueryAudienceClaimRequest {
        hash: claim_hash.clone(),
    };

    println!("1. Constructed QueryAudienceClaimRequest:");
    println!("   hash length: {} bytes", request.hash.len());
    println!("   hash (hex): {}", hex::encode(&request.hash));

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryAudienceClaimRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded hash length: {} bytes", decoded_request.hash.len());

    // === Response ===
    println!("\n--- Response ---");

    let response = QueryAudienceClaimResponse {
        claim: Some(AudienceClaim {
            signer: "xion1claimant...".to_string(),
        }),
    };

    println!("4. Constructed Response:");
    if let Some(ref claim) = response.claim {
        println!("   claim.signer: {}", claim.signer);
    }

    // 5. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response = QueryAudienceClaimResponse::decode(encoded_response.as_slice()).unwrap();
    if let Some(claim) = decoded_response.claim {
        println!("\n5. Decoded claim.signer: {}", claim.signer);
    }
}

// =============================================================================
// QueryValidateJWT
// =============================================================================
// Validate a JWT signature against a registered audience.
// Endpoint: GET /xion/jwk/validate_jwt/{aud}/{sub}/{sig_bytes}
// Type URL: /xion.jwk.v1.QueryValidateJWTRequest

fn query_validate_jwt_examples() {
    println!("\n=== QueryValidateJWT Examples ===\n");

    // 1. Construct request
    let request = QueryValidateJwtRequest {
        aud: "123456789-abc.apps.googleusercontent.com".to_string(), // Audience (OAuth client ID)
        sub: "user-unique-id-from-jwt".to_string(),                  // Subject from JWT
        sig_bytes: "base64-encoded-jwt-signature".to_string(),       // JWT signature
        ..Default::default()
    };

    println!("1. Constructed QueryValidateJwtRequest:");
    println!("   aud: {}", request.aud);
    println!("   sub: {}", request.sub);
    println!("   sig_bytes: {}", request.sig_bytes);

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryValidateJwtRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded:");
    println!("   aud: {}", decoded_request.aud);
    println!("   sub: {}", decoded_request.sub);

    // === Response ===
    println!("\n--- Response ---");

    // Response contains private claims from the validated JWT
    let response = QueryValidateJwtResponse {
        private_claims: vec![
            PrivateClaim {
                key: "email".to_string(),
                value: "user@example.com".to_string(),
            },
            PrivateClaim {
                key: "email_verified".to_string(),
                value: "true".to_string(),
            },
            PrivateClaim {
                key: "name".to_string(),
                value: "John Doe".to_string(),
            },
            PrivateClaim {
                key: "picture".to_string(),
                value: "https://example.com/photo.jpg".to_string(),
            },
        ],
    };

    println!("4. Constructed Response:");
    println!("   private_claims count: {}", response.private_claims.len());
    for claim in &response.private_claims {
        println!("     {}: {}", claim.key, claim.value);
    }

    // 5. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response = QueryValidateJwtResponse::decode(encoded_response.as_slice()).unwrap();
    println!(
        "\n5. Decoded private_claims count: {}",
        decoded_response.private_claims.len()
    );
}

// =============================================================================
// PrivateClaim Helper
// =============================================================================
// PrivateClaim represents a key-value pair from a JWT

fn private_claim_examples() {
    println!("\n=== PrivateClaim Helper Examples ===\n");

    // Construct PrivateClaims
    let claims = vec![
        PrivateClaim {
            key: "email".to_string(),
            value: "user@example.com".to_string(),
        },
        PrivateClaim {
            key: "email_verified".to_string(),
            value: "true".to_string(),
        },
        PrivateClaim {
            key: "iat".to_string(),
            value: "1699900000".to_string(), // Issued at
        },
        PrivateClaim {
            key: "exp".to_string(),
            value: "1699903600".to_string(), // Expiration
        },
    ];

    println!("1. Constructed PrivateClaims:");
    for claim in &claims {
        println!("   {}: {}", claim.key, claim.value);
    }

    // Encode/decode a single claim
    let encoded = claims[0].encode_to_vec();
    let decoded = PrivateClaim::decode(encoded.as_slice()).unwrap();
    println!("\n2. Encoded/Decoded:");
    println!("   Encoded size: {} bytes", encoded.len());
    println!("   Decoded: {} = {}", decoded.key, decoded.value);
}

// =============================================================================
// Run All Examples
// =============================================================================

fn main() {
    println!("╔════════════════════════════════════════════════════════════╗");
    println!("║  xion.jwk.v1 Query Messages — Protobuf Examples            ║");
    println!("╚════════════════════════════════════════════════════════════╝\n");

    query_params_examples();
    query_audience_examples();
    query_audience_all_examples();
    query_audience_claim_examples();
    query_validate_jwt_examples();
    private_claim_examples();

    println!("\n═══════════════════════════════════════════════════════════════");
    println!("Type URL Reference (Requests):");
    println!("  QueryParamsRequest:         /xion.jwk.v1.QueryParamsRequest");
    println!("  QueryAudienceRequest:       /xion.jwk.v1.QueryAudienceRequest");
    println!("  QueryAudienceAllRequest:    /xion.jwk.v1.QueryAudienceAllRequest");
    println!("  QueryAudienceClaimRequest:  /xion.jwk.v1.QueryAudienceClaimRequest");
    println!("  QueryValidateJWTRequest:    /xion.jwk.v1.QueryValidateJWTRequest");
    println!("\nREST Endpoints:");
    println!("  GET /xion/jwk/params");
    println!("  GET /xion/jwk/audience/{{aud}}");
    println!("  GET /xion/jwk/audience");
    println!("  GET /xion/jwk/audience_claim/{{hash}}");
    println!("  GET /xion/jwk/validate_jwt/{{aud}}/{{sub}}/{{sig_bytes}}");
    println!("═══════════════════════════════════════════════════════════════");
}
